shasta/ast_node.py

```diff
@@ -322,16 +405,18 @@ def pretty(self):
         
         if isinstance(first, NotNode):
             t = first.body
-            return f'until {t.pretty()}; do {b.pretty()}; done '
+            return f'until {t.pretty(no_braces=True) if t.NodeName == "Semi" else t.pretty()}; do \
+            {b.pretty(no_braces=True) if b.NodeName == "Semi" else b.pretty()}; done '
         else:
             t = first
-            return f'while {t.pretty()}; do {b.pretty()}; done '
+            return f'while {t.pretty(no_braces=True) if t.NodeName == "Semi" else t.pretty()}; do \
```

Comment:
It’s backward compatible in the sense of dash script still run, but the pretty printing is different. The main difference (according to Seth’s thesis paper), is that now semicolons aren’t printed when contained within several nodes since bash can’t parse those. Example: 

```bash
( { echo hi ; echo bye } )
```

This is acceptable in dash but a syntax error in bash.

Is this change relevant? I could add a check for whether shasta is parsing a bash or a dash AST and pretty prints accordingly.

Reply:
Are we still passing the dash tests? (I guess we do because they are in CI lol). That makes me feel that this is fine actually; if the change in pretty printing leads to scripts that are accepted by both we are good!